### ISSUE TITLE
Dataset.TOML with fine_tune.py does not want to work.

### ISSUE THREAD

**Issue by Hosiokaa** [kohya-ss/sd-scripts#877]
Dataset TOML threw some errors, here is the dataset.toml
It actually threw 2 errors here is the 1st one.

```
[[datasets]]
resolution = [768, 768]
min_bucket_reso = 512
max_bucket_reso = 1156

	[[datasets.subsets]]
	num_repeats = 2
	caption_extension = '.txt'
	shuffle_caption = true
	random_crop = true
	caption_dropout_rate = 0.04
	flip_aug = true
	image_dir = 'Z:\Datasets\BAKA-DIFFUSION[BLANK SMALL SCALE FINETUNE]\ConceptA'
	metadata_file = 'Z:\Datasets\BAKA-DIFFUSION[BLANK SMALL SCALE FINETUNE]\Metadata\Metadata.json'
```

```
Here is the first error :
python fine_tune.py --config "Z:\!LORA_Scripts\Vanilla Kohya\TOML template\config.toml" --dataset_config "Z:\!LORA_Scripts\Vanilla Kohya\TOML template\dataset.toml"
A matching Triton is not available, some optimizations will not be enabled.
Error caught was: No module named 'triton'
Loading settings from Z:\!LORA_Scripts\Vanilla Kohya\TOML template\config.toml...
Z:\!LORA_Scripts\Vanilla Kohya\TOML template\config
zero_terminal_snr is enabled, but v_parameterization is not enabled. training will be unexpected / zero_terminal_snrが有効ですが、v_parameterizationが有効ではありません。学習結果は想定外になる可能性があります
prepare tokenizer
update token length: 150
Load dataset config from Z:\!LORA_Scripts\Vanilla Kohya\TOML template\dataset.toml
Invalid user config / ユーザ設定の形式が正しくないようです
Traceback (most recent call last):
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\fine_tune.py", line 487, in <module>
    train(args)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\fine_tune.py", line 76, in train
    blueprint = blueprint_generator.generate(user_config, args, tokenizer=tokenizer)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\library\config_util.py", line 354, in generate
    sanitized_user_config = self.sanitizer.sanitize_user_config(user_config)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\library\config_util.py", line 318, in sanitize_user_config
    return self.user_config_validator(user_config)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\venv\lib\site-packages\voluptuous\schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\venv\lib\site-packages\voluptuous\schema_builder.py", line 595, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\venv\lib\site-packages\voluptuous\schema_builder.py", line 433, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: extra keys not allowed @ data['datasets'][0]['subsets'][0]['caption_extension']
```

It doesn't seem to accept `caption_extension = '.txt'`in either [[dataset]] [General] or [[dataset.subsets]]
This is a modified config from a LoRA TOML.

I nuked `caption_extension = '.txt'` and it now throws a different error !

```
(venv) PS Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts> python fine_tune.py --config "Z:\!LORA_Scripts\Vanilla Kohya\TOML template\config.toml" --dataset_config "Z:\!LORA_Scripts\Vanilla Kohya\TOML template\dataset.toml"
A matching Triton is not available, some optimizations will not be enabled.
Error caught was: No module named 'triton'
Loading settings from Z:\!LORA_Scripts\Vanilla Kohya\TOML template\config.toml...
Z:\!LORA_Scripts\Vanilla Kohya\TOML template\config
zero_terminal_snr is enabled, but v_parameterization is not enabled. training will be unexpected / zero_terminal_snrが有効ですが、v_parameterizationが有効ではありません。学習結果は想定外になる可能性があります
prepare tokenizer
update token length: 150
Load dataset config from Z:\!LORA_Scripts\Vanilla Kohya\TOML template\dataset.toml
loading existing metadata: Z:\Datasets\BAKA-DIFFUSION[BLANK SMALL SCALE FINETUNE]\Metadata\Metadata.json
Traceback (most recent call last):
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\fine_tune.py", line 487, in <module>
    train(args)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\fine_tune.py", line 77, in train
    train_dataset_group = config_util.generate_dataset_group_by_blueprint(blueprint.dataset_group)
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\library\config_util.py", line 430, in generate_dataset_group_by_blueprint
    dataset = dataset_klass(subsets=subsets, **asdict(dataset_blueprint.params))
  File "Z:\!LORA_Scripts\Vanilla Kohya\sd-scripts\library\train_util.py", line 1518, in __init__
    metadata = json.load(f)
  File "C:\Users\eclip\AppData\Local\Programs\Python\Python310\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "C:\Users\eclip\AppData\Local\Programs\Python\Python310\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Users\eclip\AppData\Local\Programs\Python\Python310\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Users\eclip\AppData\Local\Programs\Python\Python310\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Appears to be error for **metadata.json?**  this seems like a new option and I have no clue how to set it up. I had the idea the trainer is going to write something in there so i left it empty.

Note: I tried putting `caption_extension = '.txt'` in config.toml and it now only throws 2nd error. I'm sort of out of options here please correct me on what's wrong.  ~~(I believe trying to make this work gave me autism[🙏](https://emojipedia.org/folded-hands))~~

**Comment by DKnight54** [comment]
IIIRC, caption_extension should only be placed at the base [general] level. That might be what's causing the first error.
For the second error, have you already generated the metadata file at "Z:\Datasets\BAKA-DIFFUSION[BLANK SMALL SCALE FINETUNE]\Metadata\Metadata.json"?

If you are running from cmd line and not sure how to set it up, you can refer to this **[readme in english](https://github.com/darkstorm2150/sd-scripts/blob/main/docs/train_README-en.md#creating-metadata-files)** under the section "Creating metadata files" to prepare it first.

You'd need to use either the "merge_captions_to_metadata.py" or "merge_dd_tags_to_metadata.py" scripts to generate the metadata file before you can give the location of the file in the TOML config

**Comment by Hosiokaa** [comment]
>It doesn't seem to accept caption_extension = '.txt 'in either [[dataset]] [General] or [[dataset.subsets]]

I have tried that for some reason I still get errors.

2. I had thought that [Dataset config](https://github.com/darkstorm2150/sd-scripts/blob/main/docs/config_README-en.md) is where I would find how to setup the metadata.json thanks to you I have finally got it! 

1st is not much of an issue so I converted the captions to .caption instead of .txt which I believe is the default. but it is weird that even using an older config TOML which the `caption_extension = '.txt' ` under dataset.subsets worked didnt work on this run.

Issue is closed. Thank you very much! 
@DKnight54
